Try to follow documentation guidelines, eliminate personal pronouns (#1965)
vespa-engine/documentation: en/performance/practical-search-performance-guide.md

With `index` Vespa builds inverted index data structures which roughly consists of:

- A dictionary of the unique text tokens (after linguistic processing)
- Posting lists for each unique text token in the collection. Posting lists comes in different
formats, and using `rank: filter` can help guide the decision on what format to use. Bitvector
representation is the most compacting post list representation. 

With `attribute`, Vespa per default, does not build any inverted index like data structures for 
potential faster query evaluation. See [Wikipedia:Inverted Index](https://en.wikipedia.org/wiki/Inverted_index) 
and [Vespa internals](../proton.html#index). 
The reason for this default setting is that Vespa `attribute` fields can be used
for many different aspects: [ranking](../ranking.html), [result grouping](../grouping.html),
 [result sorting](../reference/sorting.html), and finally searching/matching. 

The following section focuses on the `tags` field which we defined with `attribute`,
matching in this field will be performed using `match:word` which is the

> default match mode for string fields with `indexing: attribute`. The
> `tags` field is as type [weightedset](../reference/schema-reference.html#type:weightedset).

Reviewer: ```suggestion
default match mode for string fields with `indexing: attribute`.
The `tags` field is of type [weightedset](../reference/schema-reference.html#type:weightedset).
```

<pre>
 field tags type weightedset&lt;string&gt; {
      indexing: summary | attribute
 }
</pre>

> Weightedset is a field type that allow representing a tag with a integer weight, which can be used for ranking. 

Reviewer: ```suggestion
Weightedset is a field type that allows representing a tag with an integer weight, which can be used for ranking. 
```

> In this case, there is no inverted index structure and matching against the `tags` field 
> is performed as a linear scan, the following scans for documents where `tags` match *rock*:

Reviewer: ```suggestion
In this case, there is no inverted index structure,
and matching against the `tags` field is performed as a linear scan.
The following scans for documents where `tags` match *rock*:
```

</pre>
</div>

> In all these examples searching the `tags` field, the matching is performed by a linear scan through all

Reviewer: ```suggestion
In all these examples searching the `tags` field, the matching is done by a linear scan through all
```

by the Vespa query planner. 

### Searching attribute fields using fast-search 
This section adds `fast-search` to the `tags` field to speed up searches where there are no 

> other query filters which restricts the search. The schema with `fast-serch`:

Reviewer: ```suggestion
other query filters which restricts the search. The schema with `fast-search`:
```

</pre>
</div>

> The above will print a `WARNING:

Reviewer: not sure if you intended markup here or not:
```suggestion
The above will print a WARNING:
```

Waiting up to 300 seconds for query service to become available ...
</pre>   

> To enable `fast-search` Vespa content node(s) needs to be restarted to re-build the fast-search data structures

Reviewer: ```suggestion
To enable `fast-search`, content node(s) needs to be restarted to re-build the fast-search data structures
```

</pre>
</div>

> This step requires to wait for the searchnode, use [health state api](../reference/metrics.html):

Reviewer: consider content node?
```suggestion
This step requires waiting for the searchnode, use the [health state api](../reference/metrics.html):
```

$ sleep 60
</pre>

> Once up, execute the `tags`query again:

Reviewer: ```suggestion
Once up, execute the `tags` query again:
```

increased memory usage and slightly reduced indexing throughput. See also
[when to use fast-search for attributes](feature-tuning.html#when-to-use-fast-search-for-attribute-fields).

For use cases requiring `match:text` when searching multi-valued string field types 
like [weightedset](../reference/schema-reference.html#type:weightedset), see
[searching multi-value fields](../searching-multi-valued-fields.html).

## Multi-valued query operators

This section covers [multi-value query operators](../multivalue-query-operators.html) 
and their query performance characteristics. Many real-world search and recommendation use cases 

> involves structured multi-valued queries. 

Reviewer: ```suggestion
involve structured multi-valued queries. 
```

}
</pre>

> The `dotProduct`and `wand` query operators produces a `rank feature` called

Reviewer: ```suggestion
The `dotProduct`and `wand` query operators produce a `rank feature` called
```

improve query performance compared to `dotProduct`. 

*wand* is a type of query operator which performs matching and ranking interleaved and skipping documents
which cannot make it into the top k results. See the [using wand with Vespa](../using-wand-with-vespa.html)
guide for more details on the WAND algorithm. 

Finally, these multi-value query operators works on both single valued fields, and array fields, 
but optimal performance is achieved using the [weightedset](../reference/schema-reference.html#type:weightedset) 
field type. The `weightedset` field type only supports integer weights. The next section
covers tensors that support more floating point number types. 

## Tensor Computations
The previous sections covered matching and where query matching query operators 

> also produced rank features which could use to influence the order of the hits returned. 

Reviewer: ```suggestion
also produced rank features which could be used to influence the order of the hits returned. 
```

In this section we look at ranking with [tensor computations](../tensor-examples.html) 
using [tensor expressions](../tensor-user-guide.html). 

Tensor computations can be used to calculate dense dot products, sparse
dot products, matrix multiplication, neural networks and more. Tensor computations can be performed 
on documents that are retrieved by the query matching operators. The only exception to this is
dense single order tensors (vectors) where Vespa also supports "matching" using [(approximate) nearest
neighbor search](..//approximate-nn-hnsw.html). 

> The `track` schema was defined a `similar` tensor field with one named *mapped* dimension. 

Reviewer: ```suggestion
The `track` schema was defined with a `similar` tensor field with one named *mapped* dimension. 
```

*Mapped* tensors can be used to represent sparse feature representations, similar

> to our `weightedset` field, but in a more generic way, and here using `float` to represent

Reviewer: ```suggestion
to the `weightedset` field, but in a more generic way, and here using `float` to represent
```

the tensor cell value. 

<pre>
field similar type tensor&lt;float&gt;(trackid{}) {
      indexing: summary | attribute
}
</pre>

> Inspecting one document, using the vespa-cli (Wraps [Vespa document/v1 api](../document-v1-api-guide.html)).

Reviewer: ```suggestion
Inspecting one document, using the vespa-cli (Wraps [Vespa document/v1 api](../document-v1-api-guide.html)):
```

</pre>
</div>

> Returns the following

Reviewer: ```suggestion
Returns:
```

}
{% endhighlight %}</pre>

Given a single track, one could just retrieve the document and display the offline computed similar tracks, but,
if a user has listened to multiple tracks in a real time session, one could use a sparse dot product
between the user recent activity and the track similarity fields. For example, listening

> to the following tracks 

Reviewer: ```suggestion
to the following tracks:
```

- `TRQIQMT128E0791D9C` Summer Of '69 by Bryan Adams
- `TRWJIPT128E0791D99` Run To You by Bryan Adams
- `TRGVORX128F4291DF1` Broken Wings by Mr. Mister

Could be represented as a query tensor `query(user_liked)` and passed with the query request like this:

<pre>

> ranking.features.query(user_liked)={{trackid:TRUAXHV128F42694E8}:1.0,{trackid:TRQIQMT128E0791D9C}:1.0,{trackid:TRGVORX128F4291DF1}:1.0}

Reviewer: might have to add raw / endraw on the pre

</pre>

> The tensor type of this query tensor was introduced in previous sections

Reviewer: ```suggestion
The tensor type of this query tensor was introduced in previous sections:
```

<pre data-test="file" data-path="app/search/query-profiles/types/root.xml">
&lt;query-profile-type id=&quot;root&quot; inherits=&quot;native&quot;&gt;
    &lt;field name=&quot;ranking.features.query(user_liked)&quot; type=&quot;tensor&amp;lt;float&amp;gt;(trackid{})&quot; /&gt;
&lt;/query-profile-type&gt;
</pre>

Both the document tensor and the query tensor are defined with `trackid{}` as the *named* *mapped* dimension. The 

> sparse tensor dot product can then be expression in a `rank-profile` like this:

Reviewer: ```suggestion
sparse tensor dot product can then be expression in a `rank-profile`:
```

}
{% endhighlight %}</pre>

> This query retrieves 95,663 document and the three tracks previously *liked*, where removed from the result.

Reviewer: ```suggestion
This query retrieves 95,663 documents, and the three tracks previously *liked* were removed from the result.
```

> The following example filters by a tags query, `tags:popular`, which reduces the complexity of the

Reviewer: ```suggestion
The following example filters by a tags query, `tags:popular`, reducing the complexity of the
```

The `querytime` dropped to 40 ms instead of 120 ms without the `fast-search` option. 
See also [performance considerations](../tensor-user-guide.html#performance-considerations)

> when using tensor expression. Vespa supports both `int8`, `bfloat16`, `float` and

Reviewer: ```suggestion
when using tensor expression. Vespa supports `int8`, `bfloat16`, `float` and
```

`double` precision cell types. A tradeoff between speed, accuracy and memory usage.

## Multi-threaded search and ranking 
So far in this guide all search queries and ranking computations have been performed using 
single threaded execution. To enable multi-threaded execution, a setting needs to be 
added to `services.xml`. 
Multi-threaded search and ranking can improve query latency significantly and make better
use of multi-cpu core architectures. Search and ranking has evolved since 1998 and by 
using multi-threaded execution, serving latency can be reduced significantly.  

> In many deployed search and recommendation systems, low serving latency is critical.  

Reviewer: ```suggestion
In many search and recommendation systems, low serving latency is critical.  
```

</pre>
</div>

And adding a new rank-profile does not require any restart, repeat the query again,
now using the `similar-t2` profile:

<div class="pre-parent">
  <button class="d-icon d-duplicate pre-copy-button" onclick="copyPreContent(this)"></button>
<pre data-test="exec" data-test-assert-contains="Bonnie Tyler">{% raw %}
$ vespa query \
    'yql=select title,artist, track_id from track where !weightedSet(track_id,@userLiked)' \
    'ranking.features.query(user_liked)={{trackid:TRQIQMT128E0791D9C}:1.0,{trackid:TRWJIPT128E0791D99}:1.0,{trackid:TRGVORX128F4291DF1}:1.0}' \
    'ranking=similar-t2' \
    'hits=5' \
    'userLiked={TRQIQMT128E0791D9C:1,TRWJIPT128E0791D99:1,TRGVORX128F4291DF1:1}'
{% endraw %}</pre>
</div>

By using multiple ranking profiles like above, developers can find the sweet-spot 
where latency does not improve much by using more threads. Using more threads per search 
limits query concurrency as more threads will be occupied
per query. Read more in [Vespa sizing guide:reduce latency with 
multi-threaded search](sizing-search.html#reduce-latency-with-multi-threaded-per-search-execution).

## Advanced range search with hitLimit  

> Vespa has an advanced query operator that allows select the 

Reviewer: ```suggestion
Vespa has an advanced query operator that allows selecting the 
```

$ vespa deploy --wait 300 app
</pre>
</div>

> Adding a new field does not require any restart, apply the partial updates by:

Reviewer: ```suggestion
Adding a new field does not require a restart, apply the partial updates by:
```

</pre>
</div>

> With that feed job completed, it is possible to use select the five tracks with the highest popularity by 

Reviewer: ```suggestion
With that feed job completed, it is possible to select the five tracks with the highest popularity by 
```

{% endraw %}</pre>
</div>

Notice that this query returns 1349 documents while the `range` search from previous example returned 1352 documents.
This is due to the `not` filter. 

The range search with `hitLimit` can be used for cases where one wants to select efficiently *top-k* of a 
single valued numeric `attribute` with `fast-search`. Some use cases which can be efficiently implemented
with it:

- Run ranking computations over the most recent documents 

> using a long to represent a timestamp (e.g, using Unix epoch).
> - Compute personalization tensor expressions over pre-selected content, e.g using popularity.

Reviewer: ```suggestion
using a `long` to represent a timestamp (e.g., using Unix epoch).
- Compute personalization tensor expressions over pre-selected content, e.g. using popularity.
```
